# Project: Winning Jeopardy

The following dataset contains 20,000 rows of the full dataset of jeopardy questions. The goal of the jeopardy analysis project is to find out if there are certain strategies that could help a person win based on the data and see any other patterns in the data

In [2]:
import pandas as pd
jeopardy= pd.read_csv('data/jeopardy.csv')
print(jeopardy.columns)

Index(['Show Number', ' Air Date', ' Round', ' Category', ' Value',
       ' Question', ' Answer'],
      dtype='object')


In [3]:
columns = []
for col in jeopardy.columns:
    col = col.strip()
    columns.append(col)
jeopardy.columns = columns
print(jeopardy.columns)
    

Index(['Show Number', 'Air Date', 'Round', 'Category', 'Value', 'Question',
       'Answer'],
      dtype='object')


In [4]:
import string
#The function normale cleans the question column so the text is normalised for analysis
def normale(stri):
    stri = stri.lower()
    for char in string.punctuation:
        stri = stri.replace(char, '')
    return stri
jeopardy['clean_question'] = jeopardy['Question'].apply(normale)

In [5]:
print(jeopardy['clean_question'].head())

0    for the last 8 years of his life galileo was u...
1    no 2 1912 olympian football star at carlisle i...
2    the city of yuma in this state has a record av...
3    in 1963 live on the art linkletter show this c...
4    signer of the dec of indep framer of the const...
Name: clean_question, dtype: object


In [6]:
#Uses same function to normalise answers
jeopardy['clean_answer'] = jeopardy['Answer'].apply(normale)
print(jeopardy['clean_answer'].head())

0    copernicus
1    jim thorpe
2       arizona
3     mcdonalds
4    john adams
Name: clean_answer, dtype: object


In [7]:
#For dollar values, takes out commas and dollar signs and converts them to integers
def normal_dollar(stri):
    stri = str(stri)
    stri = stri.replace(',', '')
    stri = stri.replace('$', '')
    if stri == 'nan' or stri == 'None':
        return 0
    else:
        stri = int(stri)
        return stri
jeopardy['clean_value'] = jeopardy['Value'].apply(normal_dollar)

print(jeopardy['clean_value'].head(10))

0    200
1    200
2    200
3    200
4    200
5    200
6    400
7    400
8    400
9    400
Name: clean_value, dtype: int64


In [8]:
jeopardy['Air Date'] = pd.to_datetime(jeopardy['Air Date'])

This checks for each row of the data, whether the answer is found inside the question. 


In [9]:
#This checks for each row of the data, whether the answer is found inside the question. 
def detect_matches(row):
    split_answer = row['clean_answer'].split()
    split_question = row['clean_question'].split()
    if "the" in split_answer:
        split_answer.remove("the")
    if len(split_answer) == 0:
        return 0
    match_count  = 0 
    for word in split_answer:
        if word in split_question:
            match_count+=1
        return match_count/len(split_answer)
jeopardy['answer_in_question'] = jeopardy.apply(detect_matches, axis = 1)


In [10]:
print(jeopardy['answer_in_question'].mean()) 
jeopardy['Air Date'].sort_values(ascending=True)

0.029624080410369708


19325   1984-09-21
19301   1984-09-21
19302   1984-09-21
19303   1984-09-21
19304   1984-09-21
           ...    
1953    2012-01-19
1954    2012-01-19
1955    2012-01-19
1945    2012-01-19
1922    2012-01-19
Name: Air Date, Length: 19999, dtype: datetime64[ns]

Around three percent of answers are inside the corresponding question. With such a low chance, it is a bad idea assume the answer will be in the question.

This compiles a set of all unique terms from the questions and also keeps track of a match count, which is seeing 
if a new word is already in the set of terms. The question overlap list takes for each row, what proportion of terms in the question are previously used terms.

In [11]:
terms_used = set()
question_overlap = []
for index, row in jeopardy.iterrows():
    split_question = row['clean_question'].split()
    split_question = [word for word in split_question if len(word)>5]
    match_count = 0
    for word in split_question:
        if word in terms_used:
            match_count+= 1
        terms_used.add(word)
    if len(split_question)>0:
        question_overlap.append(match_count/len(split_question))
    else: 
        question_overlap.append(0)
jeopardy['question_overlap'] = question_overlap 


In [12]:
print(jeopardy['question_overlap'].mean())


0.6919577992203563


This number is larger than I expected! Terms used in the questions here are lots of times found in previous questions. Based on term overlap, a good strategy is to study other important words in each question than the answer because these words and topics are likely to appear in future questions

In [13]:
import numpy as np
from random import choice
def high(row):
    if row['clean_value']>800:
        value = 1
    else: 
        value = 0
    return value
jeopardy['high_value'] = jeopardy.apply(high, axis=1)
#For each word count the number of times word is in high value vs low value question
def high_low(word):
    low_count = 0
    high_count = 0
    for index, row in jeopardy.iterrows():
        split_question = row['clean_question'].split(" ")
        if word in split_question:
            if row['high_value'] == 1:
                high_count+=1
            else:
                low_count+=1
    return high_count, low_count

terms_used_list =list(terms_used)
comparison_terms = [choice(terms_used_list) for _ in range(10)]
observed_expected = [] #High-low value pairs for each term
for term in comparison_terms:
    observed_expected.append(high_low(term))
    
observed_expected

[(2, 0),
 (0, 1),
 (1, 0),
 (1, 0),
 (33, 46),
 (0, 2),
 (1, 1),
 (1, 0),
 (1, 0),
 (0, 2)]

Summary of the above function:

1)Makes a new column in jeopardy that sets a question to high value if the value is >800 dollars

2)Creates a function that counts how many times a given word is in either a high value or low value question, returns those values in a list

3)Draws a sample of 10 words, and appends the high_low count for each word in a list

In [14]:
high_value_count = jeopardy[jeopardy['high_value'] == 1].shape[0]#shape[0] denotes length

In [15]:
low_value_count = jeopardy[jeopardy['high_value'] == 0].shape[0]

In [16]:
chi_squared = []
from scipy.stats import chisquare
import numpy as np
for obs_list in observed_expected:
    total = sum(obs_list)
    total_prop = total/jeopardy.shape[0]
    high_prop = total_prop*high_value_count #expected high values and low values of the same sample size
    low_prop = total_prop*low_value_count
    
    chisquared, pvalue = chisquare([obs_list[0], obs_list[1]], [high_prop, low_prop])
    chi_squared.append([chisquared, pvalue])

In [17]:
chi_squared

[[4.97558423439135, 0.025707519787911092],
 [0.401962846126884, 0.5260772985705469],
 [2.487792117195675, 0.11473257634454047],
 [2.487792117195675, 0.11473257634454047],
 [6.629860734564261, 0.010028310467574927],
 [0.803925692253768, 0.3699222378079571],
 [0.4448774816612795, 0.5047776487545996],
 [2.487792117195675, 0.11473257634454047],
 [2.487792117195675, 0.11473257634454047],
 [0.803925692253768, 0.3699222378079571]]

The chisquare test was performed for each of the 10 terms from the sample list. It was comparing the proportion of high and low value questions for the term with the total proportion of high and low value terms for the whole dataset to see if each terms has a significant difference from the total proportion. 

The results show that only one of the terms have a p value less than the standard 0.05, which means that high and low value questions are not associated with specific terms

I would like to do a chisquare test to see if the expected length of question for high or value questions is the same as the expected length of general questions

Steps

1) Compute the average value of high value question length and low value question length

2) Do chi2 for comparing high and low to expected value of a general question

In [19]:
high_value_list  = jeopardy[jeopardy['high_value'] == 1]
low_value_list = jeopardy[jeopardy['high_value'] == 0]
high_value_sample = [choice(list(high_value_list['Question'])) for _ in range(10)]
low_value_sample = [choice(list(low_value_list['Question'])) for _ in range(10)]


chi_low = []
question_length_average = np.mean(jeopardy['Question'].str.len())
def chi_question(sample):
    chi_high = []
    for question in sample:
      
        chisquared, pvalue = chisquare([len(question),question_length_average])
        chi_high.append([chisquared, pvalue])
    return chi_high

def chi_question(sample):
    
    chi_low= []
    for question in sample:
        
        chisquared, pvalue = chisquare([len(question),question_length_average])
        chi_low.append([chisquared, pvalue])  
        
    return chi_low

high_list = chi_question(high_value_sample)
low_list = chi_question(low_value_sample)
        



In [20]:
print("Chi square results of high value question length samples\n")
print(high_list)
print("\n")
print("Chi square results of low value question length samples\n")
print("____________")
print(low_list)

Chi square results of high value question length samples

[[2.0355581172564934, 0.15365777284612617], [0.2565407299372211, 0.6125067428219771], [58.197793197808366, 2.370448232548708e-14], [1.019797356954371, 0.3125670814661843], [5.961282396530055, 0.014623400962413519], [11.546056545872439, 0.0006789310699176715], [0.10583209255010714, 0.7449400701603923], [0.018199328389747724, 0.8926870760391505], [0.16291013646435612, 0.6864912359292312], [3.600064890654463, 0.05777731583616461]]


Chi square results of low value question length samples

____________
[[46.41881026862914, 9.549436381662367e-12], [7.40330874440334, 0.006510402096936676], [0.018199328389747724, 0.8926870760391505], [0.42544914294795244, 0.5142304754645026], [1.4732851402622678, 0.22482808702382415], [0.5249765961483579, 0.468726493568394], [1.166905436826779, 0.2800380034440265], [2.4425795362297458, 0.11808190333447399], [5.5216742613458125, 0.0187822743620027], [3.353059689274861, 0.06708022034635848]]


Majority of p values in both the high value and low value question lengths is more than 0.05. The high p values of question lengths in either the high or low value length set are not significantly different from the expected value of the question length. So, the question length is not very different between high and low value questions.

In [21]:
print(jeopardy['Category'].value_counts()[:20])

TELEVISION          51
U.S. GEOGRAPHY      50
LITERATURE          45
BEFORE & AFTER      40
AMERICAN HISTORY    40
HISTORY             40
AUTHORS             39
WORD ORIGINS        38
WORLD CAPITALS      37
BODIES OF WATER     36
SPORTS              36
SCIENCE             35
MAGAZINES           35
RHYME TIME          35
SCIENCE & NATURE    35
WORLD GEOGRAPHY     33
WORLD HISTORY       32
HISTORIC NAMES      32
ANNUAL EVENTS       32
BIRDS               31
Name: Category, dtype: int64


In [22]:
def match_count(datalist):
    terms_used = set()
    question_overlap = []
    for index, row in datalist.iterrows():
        split_question = row['clean_question'].split()
        split_question = [word for word in split_question if len(word)>5]
        match_count = 0
        for word in split_question:
            if word in terms_used:
                match_count+= 1
            terms_used.add(word)
        if len(split_question)>0:
            question_overlap.append(match_count/len(split_question))
        else: 
            question_overlap.append(0)
    return question_overlap
        


In [23]:
tele_jeopardy = jeopardy[jeopardy['Category']=='TELEVISION']
geo_jeopardy = jeopardy[jeopardy['Category']=='U.S. GEOGRAPHY']
sports_jeopardy = jeopardy[jeopardy['Category']=='SPORTS']
american_jeopardy = jeopardy[jeopardy['Category'] == 'AMERICAN HISTORY']


In [24]:
print("Repeat terms for categores:\n")

print("Geography: ")
print(np.mean(match_count(geo_jeopardy)))
print("\n")
print("Television: ")
print(np.mean(match_count(tele_jeopardy)))
print("\n")
print("Sports: ")
print(np.mean(match_count(sports_jeopardy)))
print("\n")
print("American: ")
print(np.mean(match_count(american_jeopardy)))

Repeat terms for categores:

Geography: 
0.22261904761904763


Television: 
0.2222222222222222


Sports: 
0.20634920634920637


American: 
0.1309126984126984


This final test does the repeat terms test that was done earlier in the analysis (seeing if terms in the questions were used in previous questions). Surprisingly, the per-category term use decimal is far lower than the general term use decimal (0.69). This test is also useful to see which categories have similar topics. American History has the least similar terms compared to geogrpahy, television and sports. 